logstash-core/lib/logstash/config/source/multi_local.rb

```diff
@@ -134,5 +137,28 @@ def do_warning?
       end
       !done
     end
+
+    # @param [Object] `yaml_config` the input object
+    # @return [Object] Updated object where its key etries which match the `CONFIG_KEYS_TO_WIPE_OUT_COMMENTS` do not contain comments
+    def wipeout_comments(yaml_config)
+      case yaml_config
+      when Hash
+        yaml_config.each do |key, val|
+          yaml_config[key.to_s] = wipeout_comments(val) if REMOVE_COMMENTS_CONFIG_KEYS.include?(key)
+        end
+      when Array
+        yaml_config.map { |val| wipeout_comments(val) }
+      when String
+        yaml_config.lines.map do |line|
+          if line.strip.start_with?("#")
+            ""
+          else
+            line.match?(/(?<!['"]) #/) ? line.gsub(/ (?<!['"])#.*/, '') : line
```

Comment:
I'm still learning what is acceptable under these pipeline definitions.... Would it be possible to have string literals with a `#` char inside? Perhaps something like

```
filter {
  mutate { add_field => { "test_field" => "This is real data # not a comment" } }
}
```
In that case, would we mangle this using this regex?

Reply:
it would:
```
jruby-9.4.9.0 :010 > line = 'mutate { add_field => { "test_field" => "This is real data # not a comment" } }'
 => "mutate { add_field => { \"test_field\" => \"This is real data # not a comment\" } }" 
jruby-9.4.9.0 :011 > line.match?(/(?<!['"]) #/) ? line.gsub(/ (?<!['"])#.*/, '') : line
 => "mutate { add_field => { \"test_field\" => \"This is real data" 
```

same with ruby code that may have a #:

```
jruby-9.4.9.0 :008 > line = 'filter { ruby { code => "puts # 1" } }  # test'
 => "filter { ruby { code => \"puts # 1\" } }  # test" 
jruby-9.4.9.0 :009 > line.match?(/(?<!['"]) #/) ? line.gsub(/ (?<!['"])#.*/, '') : line
 => "filter { ruby { code => \"puts"
```

Reply:
Wow! This is more than I initially thought. I have [revised a logic](https://github.com/elastic/logstash/pull/16625/commits/1615cb724bac8d0ef366b0b74fb927fa758ed650) but honestly I would avoid adding such complex art which _touches_ user's config. Maybe we document that commenting env ${VAR} is not allowed for `config.string`?

Example `pipelines.yml`:
```
- pipeline.id: "test #id1" # test
  config.string: |
    input { 
      generator { count => 10 } #id${PIPELINE_WORKERS}"
      beats {
        port => 5555 # intentional comment contains "${BEATS_DEV_PORT}" variable, shouldn't break functionalities
        host => "127.0.0.1"
      }
      # another intentional comment contains "${BEATS_PROD_HOST}" variable, shouldn't break functionalities
    }
    filter {
      mutate { add_field => { "oopsy" => "This is real data # not a ' comment, oopsy" } }
      mutate {
        add_field => {
          'doopsy' => "This is real data # not a ' comment, doopsy" # pipeline.workers: "${PIPELINE_WORKERS}"
        }
        add_tag => [ "doopsy_tag", "${ENV_TAG}" ] # "${ENV_TAG_ANOTHER}
      }
      mutate {
        add_field => {
          "hoopsy" => 
            'This is real data # not a " comment, hoopsy'
        }
      }
    mutate { add_field => { "loopsy" => "This is real data # not a ' escape comment, \\" loopsy" } }
      mutate { add_field => { "woopsy" => 'This is real data # not a " comment, ${ENV_TAG_ANOTHER} whoopsy' } } # "${ENV_TAG_ANOTHER} } }
    }
    output {
      stdout { } # output
    } # pipeline.workers: "${PIPELINE_WORKERS}"
    # pipeline.workers: "${PIPELINE_WORKERS}"

  #  pipeline.workers: "${PIPELINE_WORKERS}"

```

Reply:
Agreed, it's a very high hanging fruit for very little return in terms of user experience. +1 on documenting this limitation and closing the PR

Reply:
I agree, i was looking back at the discussion around this issue trying to get the context. It seemed to me at first glance that the behavior of doing interpolation here is correct. Specifically the flow seemed to be that variables would be interpolated before generating the config. At that point there would be no concept of what is a comment vs config in the template input. 

I do see how this may be a frustrating edge case, but it is one that is probably not worth introducing magic around.

Reply:
Thanks for the feedback! I have added doc changes with [the PR-16689](https://github.com/elastic/logstash/pull/16689) and that closes both current PR and upstream issue.